src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpResponseParser.cs

```diff
@@ -33,47 +28,56 @@ internal static class WinHttpResponseParser
             var response = new HttpResponseMessage();
             bool stripEncodingHeaders = false;
 
+            // Create a single buffer to use for all subsequent WinHttpQueryHeaders string interop calls.
+            // This buffer is the length needed for WINHTTP_QUERY_RAW_HEADERS_CRLF, which includes the status line
+            // and all headers separated by CRLF, so it should be large enough for any individual status line or header queries.
+            int bufferLength = GetResponseHeaderCharBufferLength(requestHandle, Interop.WinHttp.WINHTTP_QUERY_RAW_HEADERS_CRLF);
+            char[] buffer = new char[bufferLength];
```

Comment:
If you look at the code before this PR you'll see that it was worse before... Previously it queried `WINHTTP_QUERY_RAW_HEADERS_CRLF` along with a few other headers. Each time it did so a `StringBuilder` would be allocated (along with underlying `char[]` buffer), then there would be copying going on back and forth as part of the interop marshaling, and then a call to `ToString` would allocate a string to return, so previously there was a lot of allocations and copying going on. With this PR, all those `StringBuilder`/`char[]`/`string` allocations (and copying) are reduced to a only a single `char[]` buffer allocation.

I agree it'd be nice if the headers were lazily extracted as requested by callers -- avoiding the buffer allocation and parsing, if unnecessary.